7285 in new jupyter projects the starting notebook doesnt connect to the runtime automatically

nstrayer opened the pull request:
@:new-project-wizard

This PR fixes addresses #7285 where newly created Jupyter notebook projects didn't automatically connect to their runtime. 

When starting a new notebook project, users previously had to manually connect it to the runtime, which was confusing and unexpected. Now, the notebook automatically connects once the runtime becomes available.

The implementation uses a few key approaches:
1. Listens for runtime initialization and reacts when it's ready
2. Finds and connects the notebook to the appropriate runtime
3. Selects the best kernel based on the project's runtime
4. Cleans up event listeners once the connection is established

  ### Release Notes

  #### Bug Fixes

  - Fixed issue where new Jupyter notebook projects didn't automatically connect to their runtime

  ### QA Notes

  To test this change:
  1. Create a new Jupyter notebook project
  2. Verify the notebook automatically connects to the runtime after it has started up (no manual connection required)
  3. Confirm you can execute notebook cells without having to manually select a kernel
  4. This change affects the notebook project creation workflow, but shouldn't impact other project types

nstrayer committed:
First attempt that's slow

nstrayer committed:
Simplify and better explain logic flow

nstrayer committed:
Tiny bit of cleanup

sharon-wang reviewed (commented):
A kernel now connects automatically for my new Jupyter Notebook project 👌 

For some reason though, the new notebook always connects to a 3.10.4 Python interpreter instead of my Venv:

<img width="1681" alt="image" src="https://github.com/user-attachments/assets/79a2b0d8-214a-47a9-bdcb-2b1692342cf6" />

I tried creating a Jupyter notebook project a few times with different Python interpreters, but the notebook always chose the 3.10.4 Python instead of the created Venv. I checked my interpreter-related settings and don't see something that would cause the 3.10.4 interpreter to be selected by default.

I'm also finding that sometimes the base interpreter (the one chosen in the Wizard to create the Venv from) gets started instead of the new Venv in the console:

<img width="1063" alt="image" src="https://github.com/user-attachments/assets/3b35f069-8ecc-496a-a3e6-bd9d5d987de5" />

nstrayer commented:
Oof, good catch! That's what I get for not adding an e2e test.

nstrayer committed:
Fix bug where we were not actually selecting the right kernel for the notebooks

nstrayer committed:
Simplify by merging functions into larger easier to read ones

nstrayer committed:
- Improved the logic for selecting the active notebook editor, preferring it when available and providing fallback options.
- Added user notifications for scenarios where no notebook editors are open or when multiple editors are present without an active one.
- Updated error messages to be more user-friendly, ensuring users are informed about the need to select a kernel manually if none is found.

nstrayer committed:
Dont duplicate logging

nstrayer committed:
Add e2e test

nstrayer committed:
Make kernel finding behavior more robust on slower machines (like CI)

nstrayer commented:
Added a playwright tests. On my local machine the test fails but when I manually do the test actions it works. Pushing here to see if I just havent setup my playwright environment properly on my local machine.

nstrayer committed:
Waiting for correct kernel is now event-driven rather than polling

nstrayer committed:
Attempt to fix e2e bugs

nstrayer commented:
I fixed the selection bug on my manual runs but can't seem to get my local e2e tests running properly. I am hoping it's a setup issue on my end and CI will be able to handle it.

nstrayer committed:
Be less specific with our venv check in the kernel picker

nstrayer committed:
Reflow comment to trigger new CI run

sharon-wang commented on src/vs/workbench/services/positronNewProject/common/positronNewProjectService.ts:
```diff
@@ -13,15 +13,18 @@ import { IWorkspaceTrustManagementService } from '../../../../platform/workspace
 import { localize } from '../../../../nls.js';
-import { IRuntimeSessionService } from '../../runtimeSession/common/runtimeSessionService.js';
+import { IRuntimeSessionService, RuntimeStartMode } from '../../runtimeSession/common/runtimeSessionService.js';
+import { INotebookEditorService } from '../../../contrib/notebook/browser/services/notebookEditorService.js';
+import { INotebookKernel, INotebookKernelService } from '../../../contrib/notebook/common/notebookKernelService.js';
+import { INotebookTextModel } from '../../../contrib/notebook/common/notebookCommon.js';
```
I'm seeing a lint error on these imports. I'm not sure how best to address this though, as all the notebook services are under `workbench/contrib` instead of `editor/contrib`.

```
Bad layering. You are not allowed to access browser from here, allowed layers are: [common]eslint[local/code-layering](https://github.com/microsoft/vscode/wiki/Source-Code-Organization)
Imports violates 'vs/base/common/** or vs/base/parts/*/common/** or vs/platform/*/common/** or vs/editor/common/** or vs/editor/contrib/*/common/** or vs/workbench/common/** or vs/workbench/services/*/common/** or tas-client-umd or vscode-textmate or @vscode/vscode-languagedetection or @vscode/tree-sitter-wasm or vs/nls.js or vs/amdX.js' restrictions. See https://github.com/microsoft/vscode/wiki/Source-Code-Organizationeslint[local/code-import-patterns](https://github.com/microsoft/vscode/wiki/Source-Code-Organization)
```

sharon-wang commented on src/vs/workbench/services/positronNewProject/common/positronNewProjectService.ts:
```diff
@@ -533,6 +544,206 @@ export class PositronNewProjectService extends Disposable implements IPositronNe
+		const notebookEditors = this._notebookEditorService.listNotebookEditors();
+
+		// Prefer the active notebook editor if available
+		const activeEditor = (this._notebookEditorService as any).activeNotebookEditor as { textModel?: INotebookTextModel } | undefined;
```
Should the `INotebookEditorService` interface have `activeNotebookEditor` as a property, so that we don't need to do these type assertions?

sharon-wang reviewed (approved):
Working well for me on Mac Desktop!
